android/expoview/src/main/java/host/exp/exponent/notifications/NotificationHelper.java

```diff
@@ -116,17 +116,22 @@ public static void getPushNotificationToken(
       FcmRegistrationIntentService.getTokenAndRegister(exponentSharedPreferences.getContext());
     }
 
-    AsyncCondition.wait("devicePushToken", new AsyncCondition.AsyncConditionListener() {
+    AsyncCondition.wait(ExponentNotificationIntentService.DEVICE_PUSH_TOKEN_KEY, new AsyncCondition.AsyncConditionListener() {
       @Override
       public boolean isReady() {
-        return exponentSharedPreferences.getString(Constants.FCM_ENABLED ? ExponentSharedPreferences.FCM_TOKEN_KEY : ExponentSharedPreferences.GCM_TOKEN_KEY) != null;
+        return exponentSharedPreferences.getString(Constants.FCM_ENABLED ? ExponentSharedPreferences.FCM_TOKEN_KEY : ExponentSharedPreferences.GCM_TOKEN_KEY) != null
+            || exponentSharedPreferences.getBoolean(Constants.FCM_ENABLED ? ExponentSharedPreferences.FCM_TOKEN_ERROR_KEY : ExponentSharedPreferences.GCM_TOKEN_ERROR_KEY);
```

Comment:
Not sure we want to cache the errors -- once there's an error, `isReady()` will always return true and then we'll run `execute()` right away (looking at the AsyncCondition code). This means we'll always invoke the onFailure listener.

Also, we don't need the `updateDeviceToken` request the first time we get the native device token anymore -- it is sent to the server with the `getExpoPushToken` call.

Reply:
That's a good point (although that won't prevent the token from being written to shared preferences, so it will work on still the next run). I'll store the errored state in a static variable or something instead.